editor/gui/scene_tree_editor.cpp

```diff
@@ -646,12 +795,40 @@ void SceneTreeEditor::_update_tree(bool p_scroll_to_selected) {
 	}
 
 	updating_tree = true;
-	tree->clear();
+
 	last_hash = hash_djb2_one_64(0);
-	if (get_scene_node()) {
-		_add_nodes(get_scene_node(), nullptr);
+	Node *scene_node = get_scene_node();
+
+	if (current_scene_node != scene_node) {
+		tree->clear();
+		node_cache.clear();
+		current_scene_node = scene_node;
+	}
+
+	if (current_scene_node) {
+		_update_nodes(get_scene_node(), nullptr);
 		_compute_hash(get_scene_node(), last_hash);
+
+		// Delete any TreeItems no longer in the scene tree.
+		HashMap<Node *, CachedNode>::Iterator I = node_cache.begin();
+		while (I) {
+			if (I->value.last_epoch != node_cache_epoch) {
+				Node *key = I->key;
+				TreeItem *item = I->value.item;
+				TreeItem *parent = item->get_parent();
+				if (parent) {
+					parent->remove_child(item);
+				}
+				++I;
+				node_cache.erase(key);
```

Comment:
```suggestion
				HashMap<Node *, CachedNode>::Iterator N = I++;
				node_cache.remove(I);
				I = N
```

Reply:
Oh, if this is safe then none of this dance is necessary.

Reply:
It's not much different from what you are doing already.

Reply:
If you can do J = I, and then still refer to I after the removal then our iterators are deletable, and I should be able to just do node_cache.remove(I) and forget about this.

Reply:
That did indeed work, our HashMap iterators are just deletable :)